### ISSUE TITLE
Pdf doesn't include images

### ISSUE THREAD

**Issue by xNok** [jupyter/nbconvert#552]
When exporting to PDF it does not include images inserted via markdown

When using the markdown `![Description](pb1.png)` I got this error:
```
LaTeX Warning: File `pb1.png' not found on input line 320.

! Unable to load picture or PDF file 'pb1.png'.
<to be read again> 
                   }
l.320 \includegraphics{pb1.png}
                               
? 
! Emergency stop.
<to be read again> 
                   }
l.320 \includegraphics{pb1.png}
                               
No pages of output.
Transcript written on notebook.log.
```

When using the image html tag `<img  src=\"pb1-bfs.png\"/>` I got no image as an output

Everything is displayed correctly in the notebook for both case.

**Comment by mpacer** [comment]
Could you try converting first to `.tex` (with the `--to latex` exporter invocation) and then manually converting to pdf using `xelatex`? 

Do you run into the same error? 

We don't support pdf export from general HTML markup in markdown cells. I'm not sure if this is within or outside the purview of our intended coverage area.

**Comment by takluyver** [comment]
I think the issue here is that our PDF conversion produces latex in a temporary directory, and then runs xelatex on it there. It extracts images embedded in the notebook to a location where it can refer to them, but it doesn't copy images referenced from Markdown cells to a matching location in the temporary directory, so the references to them in the generated Latex break.

See also #52.

**Comment by xNok** [comment]
@mpacer you are right once I get the latex file i can compile it manually with my latex compiler with no issues. I guess It confirms @takluyver explanation.

Why don't we have access to `download as latex` via the file menu? Is this intentional?

**Comment by mpacer** [comment]
Great question! i'll make a PR and we'll find out.

**Comment by mpacer** [comment]
@xNok  So made a PR, but even if that is merged, I'd still recommend using the LaTeX converter from the command line because right now the download buttons aren't exactly emulating the functionality of the command line LaTeXExporter. 

Specifically, the command line LaTeX exporter will create a directory of files to represent any of your outputs that need to be called from an external source (i.e., images which are included via `\begin{figure}…\end{figure}`). In order to do this in the browser, we need to return a zip file (not just a TeX file) that will contain both the .tex and the appropriate images. Right now that zipping utility doesn't seem to be being triggered for either the LaTeX or the Markdown exporter (didn't test others but I'm guessing its the same). There are utilities for creating that zip file, but they aren't being used right now; I'm not sure why. 

This happens to be part of a more general project I'm going to be working on re: better in browser conversion capabilities. Keep an eye out for my PRs on the notebook repo in the coming weeks for updates.

**Comment by mpacer** [comment]
Correction — the LaTeX file will grab outputs and instead with download a zip if there are outputs. I just didn't realise that my test notebook for this didn't have any output images.

**Comment by jmarcoscosta** [comment]
What sintax should I use to include the images correctly into the notebooks? I've tried the latex suggestion, but it did not work...By the way, there was no command into the .tex file to include the image. (The images are correctly rendered if the notebook is opened on my browser)
I import images to my notebooks with this sintax: <img src="./handsOn1_python_figs/console.png">
If I use the image's URL instead of its path, the pdf output does not include the image as well...

**Comment by hagger98** [comment]
Please see the following code of auto generated Latex file created by Jupyter Notebook (ipynb) file. You can open the .tex extension file in the editor and go to the line 32 or 33 where you will find the bold line and comment it out. Your problem will be partially solved.
Code:
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth
\else\Gin@nat@width\fi}
\makeatother
\let\Oldincludegraphics\includegraphics
% Set max figure width to be 80% of text width, for now hardcoded.
(%)\renewcommand{\includegraphics}[1]{\Oldincludegraphics[width=.8\maxwidth]{#1}}
After commenting it out, find the image attachment command in .tex file. When you find that lines where compiler stops usually. replace the line of:
\adjustmentbox
with:
\includegraphics[scale=0.5 or whatever you desire]{yourimage.png}

Hope it solves your problem, mine is solved and i generated pdf from latex including graphics.

**Comment by xiaoleihuang** [comment]
Hi, 
I had the same issue, but I solved it by modifying the path of image file. Below is what I did:

Background: 
* Previously, I set `![ex 2.13](./image/ex2.13.png)`, the path is kind of relative path; and the notebook can not convert it to PDF

How I solved it:
* I changed the format of the path to  ![ex 2.13](image/ex2.13.png)`. Notice that I removed `./` which is to represent the current path.

My guess:
* I did not check the source code and how it works, but I guess it append the path `image/ex2.13.png` to a ``base_dir''. And if we used `./` at the head of path, it will raise errors.

**Comment by peteflorence** [comment]
In case it helps people out, a surprisingly works-great workaround for this issue is to save the notebook as .html, then "print" the .html to a .pdf.

That's:

- File --> Download as --> .html
- Open the .html
- "Print" the .html in your browser, but really save it to a .pdf

**Comment by paulperry** [comment]
@peteflorence Print Preview doesn't work for me for a full slide deck, it will only prepare a pdf of the slide being viewed. nbconvert would/should produce the full deck.

**Comment by kangwonlee** [comment]
In my case, I added a .png file using a website URL. Currently it does not work. 

Hope my comment helps.

**Comment by KuangcongLiu** [comment]
> Hi,
> I had the same issue, but I solved it by modifying the path of image file. Below is what I did:
> 
> Background:
> 
> * Previously, I set `![ex 2.13](./image/ex2.13.png)`, the path is kind of relative path; and the notebook can not convert it to PDF
> 
> How I solved it:
> 
> * I changed the format of the path to  `![ex 2.13](image/ex2.13.png)`. Notice that I removed `./` which is to represent the current path.
> 
> My guess:
> 
> * I did not check the source code and how it works, but I guess it append the path `image/ex2.13.png` to a ``base_dir''. And if we used `./` at the head of path, it will raise errors.

Thank you so much! Solved my issue by deleting "./"

**Comment by 1uciusy** [comment]
Hi, I succeed to include image when converting ipynb to pdf.

However, some funny thing showed up, the image doesn't belong where they're supposed to be.

Like, the picture I insert in the first markdown cell shows up in the next md cell or somewhere strange.

I don't quite know about latex, so I haven't consider manually edit the tex file, I wonder if there is any easy solution?

Much thanks

**Comment by t-makaro** [comment]
@1uciusy I think I have a solution: add
```
\usepackage{float}
\floatplacement{figure}{H}
```
somewhere to the top of the latex file, but after the ```\documentclass```.

Ultimately, I think that issue is worth a bug report of it's own. I'll working on submitting some of my fixes some one of my other repos back to nbconvert and this will be one of them.

**Comment by mcclaassen** [comment]
Interestingly, if the latex file to be manually converted to a pdf is created by using the JupyterLab File>Export Notebook As...   conversion. TeXworks will not find any embedded .png files. 

I advise to always create the latex files in the command line as:

> jupyter nbconvert --to latex notebook_name.ipynb

**Comment by nguyen14ck** [comment]
nbconvert --to latex
even does not detect figure place if it uses html syntax in markdown cell:
```
<img src='images/denoising.png' /> 
```

**Comment by laholmes** [comment]
> nbconvert --to latex
> even does not detect figure place if it uses html syntax in markdown cell:
> 
> ```
> <img src='images/denoising.png' /> 
> ```

see mpacer's comment on mar 20 2017 above, support for general html in markdown cells is not guaranteed

**Comment by nguyen14ck** [comment]
Thanks. 
I know but I just hope it has a place holder to let me know what I should fix in the latex file.
Currently, I work around by commenting out 
```
    /* color: #000 !important; */
```
setting wrapped text
```
{
  "MarkdownCell": {
    "cm_config": {
      "lineWrapping": true
    }
  },
  "CodeCell": {
    "cm_config": {
      "lineWrapping": true
    }
  }
}
```
and using Chrome to Print with option "Save as PDF". These work well for my job.

**Comment by SebastianSager** [comment]
Hi!

I am also using
  
jupyter nbconvert --to latex $name.ipynb

and "lose" the markup included images, e.g., 

<img width="75%" src="png/rosenbrock2.png"></img>

is not included at all in the tex document. What I did not get from the current discussion:

1. Is there another way to include an image such that there would be an automatic conversion to LaTeX?
2. Do you plan to include this in the future or not?

I use the export on a large number of lecture scripts and have some sed-based postprocessing of the LaTeX files, so I would really appreciate some way to export the file information into the tex file (even if it were commented out, then I could include it myself via sed).

Thanks a lot, best regards

**Comment by laholmes** [comment]
@SebastianSager re 1 we got image export working using matplotlib image objects

**Comment by snard6** [comment]
I fixed it by converting:
`<img src="loss.png" />`

to:
`![Loss](loss.png)`

All the comments by @mpacer and @kangwonlee are spot on and were very helpful. I'm just adding this example because I missed those on my first pass. I'm hoping this will help somebody in the future if they are scanning this issue.

**Comment by MSeal** [comment]
Some of theses issues are a real pain to fix in nbconvert. I'm planning to get some fixes in for the 6.0 release that will help with some, but no one's really spending the significant time to refactor and address some of the fundamental issues causing these (many aren't even in jupyter). If someone has some time and a will to dig into the topic I'll gladly help review / advise where I can / know what's wrong, but I and the other current maintainers are unlikely to find the time to solve all of these for some time.

**Comment by NilsNoreyson** [comment]
Hi,
I am also fighting with this issue, which, for me, roots in the relative/absolute path structures inside the provided source code. In my experience, classical file system approaches with the "file path based locating" algorithm will become more and more a problem. But alternatives a on the rise...
...Great ideas based on a somehow "contend based location system" for files, like IPFS, are a refreshing alternative to this. I would see myself interested in tackling this "Notebook - Latex - PDF - Image" issue, but I would like to know, if a IPFS based file scheme could be possibly added to Jupyter. 

Since the browser support is very good for IPFS and beyond this "technical issue", thing will become much simpler to organize and allocate. I think this could be a great investment in the portability of notebooks in general, if for instance IPFS will get included.
I'm not having the full picture in mind yet, but a good gut feeling about this.

**Comment by MSeal** [comment]
IPFS based file scheme is a pretty experimental concept, probably would be a challenge to adopt across jupyter projects as a whole. The underlying issue is with associated assets in different contexts. Filesystems have 60+ years of established precedent that we should still be trying to make usable. The main issue is with the dependent tools being so working directory constrained that they're inflexible to setting working directories / accessing relative file paths in a consistent manner. They make a lot of assumptions that don't hold up well for nbconvert's execution model.

**Comment by ghayth82** [comment]
I am facing the same issue anyone has a solution or workaround?

**Comment by alelom** [comment]
Same issue here.

**Comment by laikas123** [comment]
For anyone experiencing this issue I found what seems to be a workaround after none of the above did.

Download the notebook as an HTML file.

Move the HTML file to a directory where ALL of your pictures are and make sure you are just using either ./ or just the image name. 

For instance if you have a file "picture.png"

Either <img src = "picture.png"> or <img src = "./picture.png"> will both work.

Now assuming the HTML is in the same folder as the images and they all use the path format specified above, open the HTML in Chrome (only tested in Chrome) and the images will be there.

Press Ctrl+p and save to PDF

**Comment by kjohnsen** [comment]
This is still a problem in version 6.0.7. The workaround of converting to latex then building it myself is working fine though.

**Comment by acelere** [comment]
> I fixed it by converting:
> `<img src="loss.png" />`
> 
> to:
> `![Loss](loss.png)`
> 
> All the comments by @mpacer and @kangwonlee are spot on and were very helpful. I'm just adding this example because I missed those on my first pass. I'm hoping this will help somebody in the future if they are scanning this issue.

That did the trick for me too! Thanks for sharing!

**Comment by martinib77** [comment]
> This is still a problem in version 6.0.7. The workaround of converting to latex then building it myself is working fine though.

which command did you use to convert from latex to pdf ?

**Comment by hosseink33** [comment]
For me it was the extra `attachment` it was creating even in the latex file. Try downloading the latex file. Your image latex code should look like this

```
\begin{figure}
	\centering
	\includegraphics{YOUR_IMAGE_NAME.jpeg}
	\caption{SOME_CAPTION.jpeg}
\end{figure}
```

For me it had an extra `attachment:` which I needed to remove to make it work!

**Comment by lcoulet** [comment]
Hello, are there any news on this issue ?

**Comment by xingfanxia** [comment]
Any fix to this?

**Comment by Plinsboorg** [comment]
There are a few workarounds available:
https://github.com/jupyter/nbconvert/issues/1326

TLDR: This should work when you export to pdf
```![Alternative_text](folder_with_images/Picture.jpg)```
*I use slash (artifacts/Picture.jpg) in my path because I use Windows. In the case of Linux backslash is needed.

And this should work with webPDF export (last comment)
https://github.com/jupyter/nbconvert/issues/1079

**Comment by mattpitkin** [comment]
In my case I needed to use the html `img` tag so that I could set the size of the image using the `width` attribute, which can't be done with the pure markdown syntax. However, I found a work-around that allowed the image size to be set *and* the image would be included in the output pdf. This is based on the solution by @tashrifbillah https://github.com/jupyter/nbconvert/issues/1079#issuecomment-1012601618 and [this Stack Overflow answer](https://stackoverflow.com/a/48084050/1862861):

* display your image in its own code cell with an ipython [`Image`](https://ipython.readthedocs.io/en/stable/api/generated/IPython.display.html#IPython.display.Image) object (which allows `width` and `height` to be set):

```python
from IPython.display import display, Image
display(Image(filename="image_directory/my_image.png", width=700))
```

* edit that cell's metadata to add a tag indicating removing the cell's input (so the above code does not get displayed in the final pdf, but the "output" image does):
  * in the notebook click "View" -> "Cell toolbar" -> "Edit Metadata" (this should now show an "Edit Metadata" button above each cell in the notebook)
  * Click the "Edit Metadata" button above the required cell.
  * In the "Edit Cell Metadata" window that pops up, add the JSON:

```json
{
  "tags": [
    "remove_input"
  ]
}
```

* convert the notebook to pdf on the command line using preprocessor tags:

```bash
jupyter nbconvert --to pdf MyNotebook.ipynb --TagRemovePreprocessor.remove_input_tags remove_input
```

The pdf should now contain the image!

**Comment by normanius** [comment]
There may have been progress on that issue, see #1079. Let's hope the fix will be available in the next release of nbconvert. As of nbconvert-7.4.0, this problem was still present.

**Comment by m13v** [comment]
the trap here is the `\IfFileExists{pb1.png}{...}{}` patch people reach for to stop the Emergency stop. it swaps a loud abort that names the exact missing file for a PDF that compiles green with the figure silently dropped, so you ship a doc missing image 3 and nobody notices till print. the real break is upstream: the png never gets copied next to the generated .tex, so `\includegraphics{pb1.png}` resolves relative to nowhere. a graphicspath pointed at the notebook dir (or copying resources on export) keeps it loud when the file's actually gone.
